Add API endpoint to visualize attributes for a specific component

GiaJordan opened the pull request:
Addresses [FDS-102](https://sagebionetworks.jira.com/browse/FDS-102)
PR includes the addition of an endpoint (`/visualize/component`) to visualize the attributes for a specific component in the style of `/visualize/attributes`. Tests for this new endpoint are also included, utilizing two components.

The suffix of temporary `.jsonld` files was also adjusted to match what is being looked for in the [attributes_explorer](https://github.com/Sage-Bionetworks/schematic/blob/develop/schematic/visualization/attributes_explorer.py#L36).

[FDS-102]: https://sagebionetworks.jira.com/browse/FDS-102?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

lakikowolfe reviewed (changes requested):
Hey @GiaJordan this looks awesome! The only thing I noticed was the `Attribute` column seems to be duplicated and missing a header.

The content returned looks like this as text:
```
",Attribute,Label,Description,Required,Valid Values,Component\nrelease_scheduled,release_scheduled,ReleaseScheduled,Date that a dataset is scheduled for release,True,,DataFlow\ncontributor,contributor,Contributor,Storage project that dataset is contained in,True,,DataFlow\nComponent,Component,Component,TBD,False,,DataFlow\ndata_portal,data_portal,DataPortal,\"Has dataset been published to data portal\n\",True,\"['FALSE', 'TRUE']\",DataFlow\nembargo,embargo,Embargo,Date that indicates when a dataset is released from embargo,True,,DataFlow\ndataset_name,dataset_name,DatasetName,Unique dataset name,True,,DataFlow\nstandard_compliance,standard_compliance,StandardCompliance,Has dataset passed standard validation checks,True,\"['FALSE', 'TRUE']\",DataFlow\nreleased,released,Released,\"Has this dataset been released\n\",True,\"['FALSE', 'TRUE']\",DataFlow\nNumber of Items,Number of Items,NumberofItems,TBD,False,,DataFlow\nDataset,Dataset,Dataset,TBD,False,,DataFlow\n"
```

GiaJordan commented:
Hey @lakikowolfe thanks for checking it out. The logic for this endpoint is shared with `/visualize/attributes` so the cause of this extra column was likely from when that was implemented. As it would affect both endpoints, I've addressed it in a separate PR. addressed in #1196

GiaJordan commented:
@lakikowolfe now that @1196 is merged the extra column behavior you noted is fixed. This PR is ready for you to verify

lakikowolfe reviewed (approved):
Reviewed the output and it looks good!

GiaJordan commented on api/routes.py:
```diff
@@ -198,7 +198,7 @@ def initalize_metadata_model(schema_url):
     # retrieve a JSON-LD via URL and store it in a temporary location
     with urllib.request.urlopen(schema_url) as response:
-        with tempfile.NamedTemporaryFile(delete=False, suffix=".jsonld") as tmp_file:
+        with tempfile.NamedTemporaryFile(delete=False, suffix=".model.jsonld") as tmp_file:
```
@linglp I wanted to check on this change with you. I couldn't use the existing `/visualize/attributes` endpoint and found there were issues with the file names. Later in the code, it was trying to split the string based on `.model.jsonld` so I made this change to reflect what was being looked for. I was able to use the endpoint afterwards but wanted to review with you before merge

linglp replied:
@GiaJordan Could you double check by running `tests/test_api.py`? If all the tests passed, then it should be okay.

GiaJordan replied:
@linglp I've never been able to successfully run the api tests on my machine, even before the change. But it looks like the same tests are failing as normal

linglp replied:
You are not able to run any tests in test_api.py or only specific tests?

GiaJordan replied:
I am able to run all of them, but some consistently fail and have since they were introduced

GiaJordan replied:
Continued in [comment below](https://github.com/Sage-Bionetworks/schematic/pull/1195#issuecomment-1512174274)

linglp commented on schematic/visualization/attributes_explorer.py:
```diff
@@ -59,17 +59,55 @@ def parse_attributes(self, save_file=True):
+        else:
+            return self._parse_attributes([component], save_file, include_index)
+
+    def _parse_attributes(self, components, save_file=True, include_index=True):
```
Could you add tests for this function: `_parase_attributes?`

GiaJordan replied:
@linglp this function is covered by tests `test_visualize_component` and `test_visualize_attributes` that are in `test_api.py::TestSchemaVisualization`

linglp replied:
@GiaJordan But these two tests only ensure that response code are 200? I think we should still add tests more specifically for ensuring that parse attributes return expected output?

GiaJordan replied:
Also, the content of `_parse_attributes` is was just pulled from `parse_attributes` into its own function so that it could be called from another function as well, with the added flag

GiaJordan commented on schematic/visualization/attributes_explorer.py:
```diff
@@ -59,17 +59,55 @@ def parse_attributes(self, save_file=True):
         # have to provide.
+        return self._parse_attributes(components, save_file)
+    
+    def parse_component_attributes(self, component=None, save_file=True, include_index=True):
```
@linglp the test is what was added in `test_api.py` highlighted in this PR

linglp replied:
@GiaJordan No. I think the test in test_api.py is for ensuring that the api endpoint works as expected. For example, when generating a manifest, the API supports generating multiple manifests for different data types. The test for the APIs are the test that ensure that this layer of logic added in the API could work as expected. But this is not the same for testing the function of generating a manifest. 

I think it is the same case here: the test that you have in test_api.py ensures that the response code is 200. But it doesn't ensure that the function "parse_component_attributes" returns expected result.

GiaJordan replied:
I see, I've added the statements to check the output of the this endpoint and ensure that the correct component information is returned

mialy-defelice replied:
Sorry to add to this @GiaJordan but I wonder if an addition to `test_viz.py` would also be appropriate?

GiaJordan replied:
@mialy-defelice I think it would be, I can add

GiaJordan replied:
@mialy-defelice I've added the test `test_ce` to the viz test suite below

linglp commented on schematic/visualization/attributes_explorer.py:
```diff
@@ -59,17 +59,55 @@ def parse_attributes(self, save_file=True):
+                False: merged_df is returned.
+              components (list):
+                list of components to parse attributes for
+
```
Could you add explanation of "include_index" since it is now part of the arguments?

GiaJordan replied:
Yes, added!

GiaJordan commented:
Tests have been added in `test_viz.py` to cover the functions that enable the functionality of the endpoint added in the PR, and in `test_api.py` to cover the functionality of the endpoint itself and the output of the endpoint.
@mialy-defelice @linglp

linglp commented on tests/test_api.py:
```diff
@@ -714,6 +714,23 @@ def test_visualize_tangled_tree_layers(self, client, figure_type, data_model_jso
 
         assert response.status_code == 200
 
+    @pytest.mark.parametrize("component", ["Patient", "BulkRNA-seqAssay"])
```
@GiaJordan could you use parameterize based on the example here: https://docs.pytest.org/en/6.2.x/parametrize.html#:~:text=42)%5D)%0Adef-,test_eval,-(test_input%2C I think that will simplify the code

GiaJordan replied:
Sure @linglp, I can parameterize the expected responses as well

GiaJordan commented:
I do not believe the renaming of the temporary file suffix will be troublesome, as it does not change the result of any of the api unit tests @milen-sage @linglp. Details below

Using GH codespaces as suggested with this branch revealed the following api test failures:
```FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest[table_and_file-true-None] - assert 500 == 200
FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest[table_and_file-false-None] - assert 500 == 200
FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_w_entities[table_file_and_entities-None] - assert 500 == 200
FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_upsert[None] - assert 400 == 200
FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_upsert[[{ "Component": "MockRDB", "MockRDB_id": 5 }]] - assert 400 == 200
FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check URL] - assert 86.36419494300208 < 5.0
FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match at Least] - assert 10.76112667200141 < 5.0
FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match at Least values] - assert 10.473595325001952 < 5.0
FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match Exactly] - assert 10.71795495599872 < 5.0
FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match Exactly values] - assert 10.466644798998459 < 5.0
```

These are the same test failures that I encountered running the api tests on `develop`:

```
FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest[table_and_file-true-None] - assert 500 == 200
FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest[table_and_file-false-None] - assert 500 == 200
FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_w_entities[table_file_and_entities-None] - assert 500 == 200
FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_upsert[None] - assert 400 == 200
FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_upsert[[{ "Component": "MockRDB", "MockRDB_id": 5 }]] - assert 400 == 200
FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check URL] - assert 92.77976941500128 < 5.0
FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match at Least] - assert 11.06076004299939 < 5.0
FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match at Least values] - assert 10.112847790998785 < 5.0
FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match Exactly] - assert 9.983342362000258 < 5.0
FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match Exactly values] - assert 9.87761154799955 < 5.0
```

linglp commented:
> I do not believe the renaming of the temporary file suffix will be troublesome, as it does not change the result of any of the api unit tests @milen-sage @linglp. Details below
> 
> Using GH codespaces as suggested with this branch revealed the following api test failures:
> 
> ```
> FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest[table_and_file-true-None] - assert 500 == 200
> FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest[table_and_file-false-None] - assert 500 == 200
> FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_w_entities[table_file_and_entities-None] - assert 500 == 200
> FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_upsert[None] - assert 400 == 200
> FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_upsert[[{ "Component": "MockRDB", "MockRDB_id": 5 }]] - assert 400 == 200
> FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check URL] - assert 86.36419494300208 < 5.0
> FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match at Least] - assert 10.76112667200141 < 5.0
> FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match at Least values] - assert 10.473595325001952 < 5.0
> FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match Exactly] - assert 10.71795495599872 < 5.0
> FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match Exactly values] - assert 10.466644798998459 < 5.0
> ```
> 
> These are the same test failures that I encountered running the api tests on `develop`:
> 
> ```
> FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest[table_and_file-true-None] - assert 500 == 200
> FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest[table_and_file-false-None] - assert 500 == 200
> FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_w_entities[table_file_and_entities-None] - assert 500 == 200
> FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_upsert[None] - assert 400 == 200
> FAILED tests/test_api.py::TestManifestOperation::test_submit_manifest_upsert[[{ "Component": "MockRDB", "MockRDB_id": 5 }]] - assert 400 == 200
> FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check URL] - assert 92.77976941500128 < 5.0
> FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match at Least] - assert 11.06076004299939 < 5.0
> FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match at Least values] - assert 10.112847790998785 < 5.0
> FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match Exactly] - assert 9.983342362000258 < 5.0
> FAILED tests/test_api.py::TestValidationBenchmark::test_validation_performance[Check Match Exactly values] - assert 9.87761154799955 < 5.0
> ```

I am not worried about the validation performance test, but TestManifestOperation::test_submit_manifest failing is concerning. Could you create a separate issue on Jira so that we could look into the test failing in another PR?

GiaJordan commented:
> I am not worried about the validation performance test, but TestManifestOperation::test_submit_manifest failing is concerning. Could you create a separate issue on Jira so that we could look into the test failing in another PR?

Sure @linglp it's been created at FDS-318
